File: airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/java/io/airbyte/cdk/extensions/LoggingInvocationInterceptor.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cdk.extensions;

import static java.util.concurrent.TimeUnit.DAYS;
import static java.util.concurrent.TimeUnit.HOURS;
import static java.util.concurrent.TimeUnit.MICROSECONDS;
import static java.util.concurrent.TimeUnit.MILLISECONDS;
import static java.util.concurrent.TimeUnit.MINUTES;
import static java.util.concurrent.TimeUnit.NANOSECONDS;
import static java.util.concurrent.TimeUnit.SECONDS;
import static java.util.regex.Pattern.CASE_INSENSITIVE;
import static java.util.regex.Pattern.UNICODE_CASE;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.lang.reflect.Proxy;
import java.time.Duration;
import java.time.Instant;
import java.time.format.DateTimeParseException;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.exception.ExceptionUtils;
import org.apache.commons.lang3.time.DurationFormatUtils;
import org.junit.jupiter.api.Timeout;
import org.junit.jupiter.api.extension.DynamicTestInvocationContext;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.api.extension.InvocationInterceptor;
import org.junit.jupiter.api.extension.ReflectiveInvocationContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * By default, junit only output logs to the console, and nothing makes it into log4j logs. This
 * class fixes that by using the interceptor facility to print progress and timing information. This
 * allows us to have junit loglines in our test logs. This is instanciated via <a href=
 * "https://docs.oracle.com/javase%2F9%2Fdocs%2Fapi%2F%2F/java/util/ServiceLoader.html">Java's
 * ServiceLoader</a> The declaration can be found in
 * resources/META-INF/services/org.junit.jupiter.api.extension.Extension
 */
public class LoggingInvocationInterceptor implements InvocationInterceptor {

  private static final Logger LOGGER = LoggerFactory.getLogger(LoggingInvocationInterceptor.class);
  private static final String JUNIT_METHOD_EXECUTION_TIMEOUT_PROPERTY_NAME = "JunitMethodExecutionTimeout";

  private static final class LoggingInvocationInterceptorHandler implements InvocationHandler {

    private static final Pattern methodPattern = Pattern.compile("intercept(.*)Method");

    @Override
    @SuppressWarnings("unchecked")
    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
      if (LoggingInvocationInterceptor.class.getDeclaredMethod(method.getName(), Invocation.class, ReflectiveInvocationContext.class,
          ExtensionContext.class) == null) {
        LOGGER.error("Junit LoggingInvocationInterceptor executing unknown interception point {}", method.getName());
        return method.invoke(proxy, args);
      }
      var invocation = (Invocation<?>) args[0];
      var invocationContext = (ReflectiveInvocationContext<Method>) args[1];
      var extensionContext = (ExtensionContext) args[2];
      String methodName = method.getName();
      String logLineSuffix;
      Matcher methodMatcher = methodPattern.matcher(methodName);
      if (methodName.equals("interceptDynamicTest")) {
        logLineSuffix = "execution of DynamicTest %s".formatted(extensionContext.getDisplayName());
      } else if (methodName.equals("interceptTestClassConstructor")) {
        logLineSuffix = "instance creation for %s".formatted(invocationContext.getTargetClass());
      } else if (methodMatcher.matches()) {
        String interceptedEvent = methodMatcher.group(1);
        logLineSuffix = "execution of @%s method %s.%s".formatted(interceptedEvent,
            invocationContext.getExecutable().getDeclaringClass().getSimpleName(),
            invocationContext.getExecutable().getName());
      } else {
        logLineSuffix = "execution of unknown intercepted call %s".formatted(methodName);
      }
      Thread currentThread = Thread.currentThread();
      TimeoutInteruptor timeoutTask = new TimeoutInteruptor(currentThread);
      Instant start = Instant.now();
      try {
        final Object retVal;
        Duration timeout = getTimeout(invocationContext);
        if (timeout != null) {
          LOGGER.info("Junit starting {} with timeout of {}", logLineSuffix, DurationFormatUtils.formatDurationWords(timeout.toMillis(), true, true));
          new Timer("TimeoutTimer-" + currentThread.getName(), true).schedule(timeoutTask, timeout.toMillis());
        } else {
          LOGGER.warn("Junit starting {} with no timeout", logLineSuffix);
        }
        retVal = invocation.proceed();
        long elapsedMs = Duration.between(start, Instant.now()).toMillis();
        LOGGER.info("Junit completed {} in {}", logLineSuffix, DurationFormatUtils.formatDurationWords(elapsedMs, true, true));
        return retVal;
      } catch (Throwable t) {
        timeoutTask.cancel();
        long elapsedMs = Duration.between(start, Instant.now()).toMillis();
        if (timeoutTask.wasTriggered) {
          Throwable t1 = t;
          t = new TimeoutException(
              "Execution was cancelled after %s. If you think your test should be given more time to complete, you can use the @Timeout annotation. If all the test of a connector are slow, "
                  + " you can override the property 'JunitMethodExecutionTimeout' in your gradle.properties."
                      .formatted(DurationFormatUtils.formatDurationWords(elapsedMs, true, true)));
          t.initCause(t1);
        }
        boolean belowCurrentCall = false;
        List<String> stackToDisplay = new LinkedList<>();
        for (String stackString : ExceptionUtils.getStackFrames(t)) {
          if (stackString.startsWith("\tat ")) {
            if (!belowCurrentCall && stackString.contains(LoggingInvocationInterceptor.class.getCanonicalName())) {
              belowCurrentCall = true;
            }
          } else {
            belowCurrentCall = false;
          }
          if (!belowCurrentCall) {
            stackToDisplay.add(stackString);
          }
        }
        String stackTrace = StringUtils.join(stackToDisplay, "\n    ");
        LOGGER.error("Junit exception throw during {} after {}:\n{}", logLineSuffix, DurationFormatUtils.formatDurationWords(elapsedMs, true, true),
            stackTrace);
        throw t;
      } finally {
        timeoutTask.cancel();
      }
    }

    private static class TimeoutInteruptor extends TimerTask {

      private final Thread parentThread;
      volatile boolean wasTriggered = false;

      TimeoutInteruptor(Thread parentThread) {
        this.parentThread = parentThread;
      }

      @Override
      public void run() {
        wasTriggered = true;
        parentThread.interrupt();
      }

      public boolean cancel() {
        return super.cancel();
      }

    }

    private static final Pattern PATTERN = Pattern.compile("([1-9]\\d*) *((?:[nμm]?s)|m|h|d)?",
        CASE_INSENSITIVE | UNICODE_CASE);
    private static final Map<String, TimeUnit> UNITS_BY_ABBREVIATION;

    static {
      Map<String, TimeUnit> unitsByAbbreviation = new HashMap<>();
      unitsByAbbreviation.put("ns", NANOSECONDS);
      unitsByAbbreviation.put("μs", MICROSECONDS);
      unitsByAbbreviation.put("ms", MILLISECONDS);
      unitsByAbbreviation.put("s", SECONDS);
      unitsByAbbreviation.put("m", MINUTES);
      unitsByAbbreviation.put("h", HOURS);
      unitsByAbbreviation.put("d", DAYS);
      UNITS_BY_ABBREVIATION = Collections.unmodifiableMap(unitsByAbbreviation);
    }

    static Duration parseDuration(String text) throws DateTimeParseException {
      Matcher matcher = PATTERN.matcher(text.trim());
      if (matcher.matches()) {
        long value = Long.parseLong(matcher.group(1));
        String unitAbbreviation = matcher.group(2);
        TimeUnit unit = unitAbbreviation == null ? SECONDS
            : UNITS_BY_ABBREVIATION.get(unitAbbreviation.toLowerCase(Locale.ENGLISH));
        return Duration.ofSeconds(unit.toSeconds(value));
      }
      throw new DateTimeParseException("Timeout duration is not in the expected format (<number> [ns|μs|ms|s|m|h|d])",
          text, 0);
    }

    private static Duration getTimeout(ReflectiveInvocationContext<Method> invocationContext) {
      Duration timeout = null;
      if (invocationContext.getExecutable()instanceof Method m) {
        Timeout timeoutAnnotation = m.getAnnotation(Timeout.class);
        if (timeoutAnnotation == null) {
          timeoutAnnotation = invocationContext.getTargetClass().getAnnotation(Timeout.class);
        }
        if (timeoutAnnotation != null) {
          timeout = Duration.ofMillis(timeoutAnnotation.unit().toMillis(timeoutAnnotation.value()));
        }
      }
      if (timeout == null) {
        timeout = parseDuration(System.getProperty(JUNIT_METHOD_EXECUTION_TIMEOUT_PROPERTY_NAME));
      }
      return timeout;
    }

  }

  private final InvocationInterceptor proxy = (InvocationInterceptor) Proxy.newProxyInstance(
      getClass().getClassLoader(),
      new Class[] {InvocationInterceptor.class},
      new LoggingInvocationInterceptorHandler());

  @Override
  public void interceptAfterAllMethod(Invocation<Void> invocation,
                                      ReflectiveInvocationContext<Method> invocationContext,
                                      ExtensionContext extensionContext)
      throws Throwable {
    proxy.interceptAfterAllMethod(invocation, invocationContext, extensionContext);
  }

  @Override
  public void interceptAfterEachMethod(Invocation<Void> invocation,
                                       ReflectiveInvocationContext<Method> invocationContext,
                                       ExtensionContext extensionContext)
      throws Throwable {
    proxy.interceptAfterEachMethod(invocation, invocationContext, extensionContext);
  }

  @Override
  public void interceptBeforeAllMethod(Invocation<Void> invocation,
                                       ReflectiveInvocationContext<Method> invocationContext,
                                       ExtensionContext extensionContext)
      throws Throwable {
    proxy.interceptBeforeAllMethod(invocation, invocationContext, extensionContext);
  }

  @Override
  public void interceptBeforeEachMethod(Invocation<Void> invocation,
                                        ReflectiveInvocationContext<Method> invocationContext,
                                        ExtensionContext extensionContext)
      throws Throwable {
    proxy.interceptBeforeEachMethod(invocation, invocationContext, extensionContext);
  }

  @Override
  public void interceptDynamicTest(Invocation<Void> invocation,
                                   DynamicTestInvocationContext invocationContext,
                                   ExtensionContext extensionContext)
      throws Throwable {
    proxy.interceptDynamicTest(invocation, invocationContext, extensionContext);
  }

  @Override
  public void interceptTestMethod(Invocation<Void> invocation,
                                  ReflectiveInvocationContext<Method> invocationContext,
                                  ExtensionContext extensionContext)
      throws Throwable {
    if (!Modifier.isPublic(invocationContext.getExecutable().getModifiers())) {
      LOGGER.warn("Junit method {}.{} is not declared as public", invocationContext.getExecutable().getDeclaringClass().getCanonicalName(),
          invocationContext.getExecutable().getName());
    }
    proxy.interceptTestMethod(invocation, invocationContext, extensionContext);
  }

  @Override
  public void interceptTestTemplateMethod(Invocation<Void> invocation,
                                          ReflectiveInvocationContext<Method> invocationContext,
                                          ExtensionContext extensionContext)
      throws Throwable {
    proxy.interceptTestTemplateMethod(invocation, invocationContext, extensionContext);
  }

  @Override
  public <T> T interceptTestFactoryMethod(Invocation<T> invocation,
                                          ReflectiveInvocationContext<Method> invocationContext,
                                          ExtensionContext extensionContext)
      throws Throwable {
    return proxy.interceptTestFactoryMethod(invocation, invocationContext, extensionContext);
  }

  @Override
  public <T> T interceptTestClassConstructor(Invocation<T> invocation,
                                             ReflectiveInvocationContext<Constructor<T>> invocationContext,
                                             ExtensionContext extensionContext)
      throws Throwable {
    return proxy.interceptTestClassConstructor(invocation, invocationContext, extensionContext);
  }

}

File: airbyte-cdk/java/airbyte-cdk/core/src/testFixtures/kotlin/io/airbyte/cdk/extensions/LoggingInvocationInterceptor.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.extensions

import io.github.oshai.kotlinlogging.KotlinLogging
import java.lang.reflect.*
import java.time.Duration
import java.time.Instant
import java.time.format.DateTimeParseException
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import java.util.regex.Pattern
import kotlin.concurrent.Volatile
import org.apache.commons.lang3.StringUtils
import org.apache.commons.lang3.exception.ExceptionUtils
import org.apache.commons.lang3.time.DurationFormatUtils
import org.junit.jupiter.api.Timeout
import org.junit.jupiter.api.extension.DynamicTestInvocationContext
import org.junit.jupiter.api.extension.ExtensionContext
import org.junit.jupiter.api.extension.InvocationInterceptor
import org.junit.jupiter.api.extension.ReflectiveInvocationContext

private val LOGGER = KotlinLogging.logger {}
/**
 * By default, junit only output logs to the console, and nothing makes it into log4j logs. This
 * class fixes that by using the interceptor facility to print progress and timing information. This
 * allows us to have junit loglines in our test logs. This is instanciated via
 * [Java's
 * ServiceLoader](https://docs.oracle.com/javase%2F9%2Fdocs%2Fapi%2F%2F/java/util/ServiceLoader.html)
 * The declaration can be found in
 * resources/META-INF/services/org.junit.jupiter.api.extension.Extension
 */
class LoggingInvocationInterceptor : InvocationInterceptor {
    private class LoggingInvocationInterceptorHandler : InvocationHandler {
        @Throws(Throwable::class)
        override fun invoke(proxy: Any, method: Method, args: Array<Any>): Any? {
            val methodName = method.name
            val invocationContextClass: Class<*> =
                when (methodName) {
                    "interceptDynamicTest" -> DynamicTestInvocationContext::class.java
                    else -> ReflectiveInvocationContext::class.java
                }
            try {
                LoggingInvocationInterceptor::class
                    .java
                    .getDeclaredMethod(
                        method.name,
                        InvocationInterceptor.Invocation::class.java,
                        invocationContextClass,
                        ExtensionContext::class.java
                    )
            } catch (_: NoSuchMethodException) {
                LOGGER.error(
                    "Junit LoggingInvocationInterceptor executing unknown interception point {}",
                    method.name
                )
                return method.invoke(proxy, *(args))
            }
            val invocation = args[0] as InvocationInterceptor.Invocation<*>?
            val reflectiveInvocationContext = args[1] as? ReflectiveInvocationContext<*>
            val extensionContext = args[2] as ExtensionContext?
            val logLineSuffix: String
            val methodMatcher = methodPattern.matcher(methodName)
            if (methodName == "interceptDynamicTest") {
                logLineSuffix = "execution of DynamicTest ${extensionContext!!.displayName}"
            } else if (methodName == "interceptTestClassConstructor") {
                logLineSuffix =
                    "instance creation for %s".formatted(reflectiveInvocationContext!!.targetClass)
            } else if (methodMatcher.matches()) {
                val interceptedEvent = methodMatcher.group(1)
                val methodRealClassName =
                    reflectiveInvocationContext!!.executable!!.declaringClass.simpleName
                val methodName = reflectiveInvocationContext.executable!!.name
                val targetClassName = reflectiveInvocationContext.targetClass.simpleName
                val methodDisplayName =
                    if (targetClassName == methodRealClassName) methodName
                    else "$methodName($methodRealClassName)"
                logLineSuffix =
                    "execution of @%s method %s.%s".formatted(
                        interceptedEvent,
                        targetClassName,
                        methodDisplayName
                    )
                TestContext.CURRENT_TEST_NAME.set("$targetClassName.$methodName")
            } else {
                logLineSuffix = "execution of unknown intercepted call $methodName"
            }
            val currentThread = Thread.currentThread()
            val timeoutTask = TimeoutInteruptor(currentThread)
            val start = Instant.now()
            try {
                val timeout = reflectiveInvocationContext?.let(::getTimeout)
                if (timeout != null) {
                    LOGGER.info(
                        "Junit starting {} with timeout of {}",
                        logLineSuffix,
                        DurationFormatUtils.formatDurationWords(timeout.toMillis(), true, true)
                    )
                    Timer("TimeoutTimer-" + currentThread.name, true)
                        .schedule(timeoutTask, timeout.toMillis())
                } else {
                    LOGGER.warn("Junit starting {} with no timeout", logLineSuffix)
                }
                val retVal = invocation!!.proceed()
                val elapsedMs = Duration.between(start, Instant.now()).toMillis()
                LOGGER.info(
                    "Junit completed {} in {}",
                    logLineSuffix,
                    DurationFormatUtils.formatDurationWords(elapsedMs, true, true)
                )
                return retVal
            } catch (throwable: Throwable) {
                timeoutTask.cancel()
                val elapsedMs = Duration.between(start, Instant.now()).toMillis()
                val t1: Throwable
                if (timeoutTask.wasTriggered) {
                    val timeoutAsString =
                        DurationFormatUtils.formatDurationWords(elapsedMs, true, true)
                    t1 =
                        TimeoutException(
                            "Execution was cancelled after $timeoutAsString. If you think your test should be given more time to complete, " +
                                "you can use the @Timeout annotation. If all the test of a connector are slow, " +
                                " you can override the property 'JunitMethodExecutionTimeout' in your gradle.properties."
                        )
                    t1.initCause(throwable)
                } else {
                    t1 = throwable
                }
                var belowCurrentCall = false
                val stackToDisplay: MutableList<String> = LinkedList()
                for (stackString in ExceptionUtils.getStackFrames(throwable)) {
                    if (stackString!!.startsWith("\tat ")) {
                        if (
                            !belowCurrentCall &&
                                stackString.contains(
                                    LoggingInvocationInterceptor::class.java.canonicalName
                                )
                        ) {
                            belowCurrentCall = true
                        }
                    } else {
                        belowCurrentCall = false
                    }
                    if (!belowCurrentCall) {
                        stackToDisplay.add(stackString)
                    }
                }
                val stackTrace = StringUtils.join(stackToDisplay, "\n    ")
                LOGGER.error(
                    "Junit exception throw during {} after {}:\n{}",
                    logLineSuffix,
                    DurationFormatUtils.formatDurationWords(elapsedMs, true, true),
                    stackTrace
                )
                throw t1
            } finally {
                timeoutTask.cancel()
                TestContext.CURRENT_TEST_NAME.set(null)
            }
        }

        private class TimeoutInteruptor(private val parentThread: Thread) : TimerTask() {
            @Volatile var wasTriggered: Boolean = false

            override fun run() {
                LOGGER.info(
                    "interrupting running task on ${parentThread.name}. Current Stacktrace is ${parentThread.stackTrace.asList()}"
                )
                wasTriggered = true
                parentThread.interrupt()
            }

            override fun cancel(): Boolean {
                LOGGER.info("cancelling timer task on ${parentThread.name}")
                return super.cancel()
            }
        }

        companion object {
            private val methodPattern: Pattern = Pattern.compile("intercept(.*)Method")

            private val PATTERN: Pattern =
                Pattern.compile(
                    "([1-9]\\d*) *((?:[nμm]?s)|m|h|d)?",
                    Pattern.CASE_INSENSITIVE or Pattern.UNICODE_CASE
                )
            private val UNITS_BY_ABBREVIATION: MutableMap<String, TimeUnit>

            init {
                val unitsByAbbreviation: MutableMap<String, TimeUnit> = HashMap()
                unitsByAbbreviation["ns"] = TimeUnit.NANOSECONDS
                unitsByAbbreviation["μs"] = TimeUnit.MICROSECONDS
                unitsByAbbreviation["ms"] = TimeUnit.MILLISECONDS
                unitsByAbbreviation["s"] = TimeUnit.SECONDS
                unitsByAbbreviation["m"] = TimeUnit.MINUTES
                unitsByAbbreviation["h"] = TimeUnit.HOURS
                unitsByAbbreviation["d"] = TimeUnit.DAYS
                UNITS_BY_ABBREVIATION = Collections.unmodifiableMap(unitsByAbbreviation)
            }

            @Throws(DateTimeParseException::class)
            fun parseDuration(text: String): Duration {
                val matcher = PATTERN.matcher(text.trim { it <= ' ' })
                if (matcher.matches()) {
                    val value = matcher.group(1).toLong()
                    val unitAbbreviation = matcher.group(2)
                    val unit =
                        if (unitAbbreviation == null) TimeUnit.SECONDS
                        else UNITS_BY_ABBREVIATION.getValue(unitAbbreviation.lowercase())
                    return Duration.ofSeconds(unit.toSeconds(value))
                }
                throw DateTimeParseException(
                    "Timeout duration is not in the expected format (<number> [ns|μs|ms|s|m|h|d])",
                    text,
                    0
                )
            }

            private fun getTimeout(invocationContext: ReflectiveInvocationContext<*>): Duration {
                var timeout: Duration? = null
                var m = invocationContext.executable
                if (m is Method) {
                    var timeoutAnnotation: Timeout? = m.getAnnotation(Timeout::class.java)
                    if (timeoutAnnotation == null) {
                        timeoutAnnotation =
                            invocationContext.targetClass.getAnnotation(Timeout::class.java)
                    }
                    if (timeoutAnnotation != null) {
                        timeout =
                            Duration.ofMillis(
                                timeoutAnnotation.unit.toMillis(timeoutAnnotation.value)
                            )
                    }
                }
                if (timeout == null) {
                    timeout =
                        parseDuration(
                            System.getProperty(JUNIT_METHOD_EXECUTION_TIMEOUT_PROPERTY_NAME)
                        )
                }
                return timeout
            }
        }
    }

    private val proxy: InvocationInterceptor? =
        Proxy.newProxyInstance(
            javaClass.classLoader,
            arrayOf<Class<*>?>(InvocationInterceptor::class.java),
            LoggingInvocationInterceptorHandler()
        ) as InvocationInterceptor

    @Throws(Throwable::class)
    override fun interceptAfterAllMethod(
        invocation: InvocationInterceptor.Invocation<Void>?,
        invocationContext: ReflectiveInvocationContext<Method>?,
        extensionContext: ExtensionContext?
    ) {
        proxy!!.interceptAfterAllMethod(invocation, invocationContext, extensionContext)
    }

    @Throws(Throwable::class)
    override fun interceptAfterEachMethod(
        invocation: InvocationInterceptor.Invocation<Void>?,
        invocationContext: ReflectiveInvocationContext<Method>?,
        extensionContext: ExtensionContext?
    ) {
        proxy!!.interceptAfterEachMethod(invocation, invocationContext, extensionContext)
    }

    @Throws(Throwable::class)
    override fun interceptBeforeAllMethod(
        invocation: InvocationInterceptor.Invocation<Void>?,
        invocationContext: ReflectiveInvocationContext<Method>?,
        extensionContext: ExtensionContext?
    ) {
        proxy!!.interceptBeforeAllMethod(invocation, invocationContext, extensionContext)
    }

    @Throws(Throwable::class)
    override fun interceptBeforeEachMethod(
        invocation: InvocationInterceptor.Invocation<Void>?,
        invocationContext: ReflectiveInvocationContext<Method>?,
        extensionContext: ExtensionContext?
    ) {
        proxy!!.interceptBeforeEachMethod(invocation, invocationContext, extensionContext)
    }

    @Throws(Throwable::class)
    override fun interceptDynamicTest(
        invocation: InvocationInterceptor.Invocation<Void>?,
        invocationContext: DynamicTestInvocationContext?,
        extensionContext: ExtensionContext?
    ) {
        proxy!!.interceptDynamicTest(invocation, invocationContext, extensionContext)
    }

    @Throws(Throwable::class)
    override fun interceptTestMethod(
        invocation: InvocationInterceptor.Invocation<Void>,
        invocationContext: ReflectiveInvocationContext<Method>,
        extensionContext: ExtensionContext
    ) {
        if (!Modifier.isPublic(invocationContext.executable!!.modifiers)) {
            LOGGER!!.warn(
                "Junit method {}.{} is not declared as public",
                invocationContext.executable!!.declaringClass.canonicalName,
                invocationContext.executable!!.name
            )
        }
        proxy!!.interceptTestMethod(invocation, invocationContext, extensionContext)
    }

    @Throws(Throwable::class)
    override fun interceptTestTemplateMethod(
        invocation: InvocationInterceptor.Invocation<Void>?,
        invocationContext: ReflectiveInvocationContext<Method>?,
        extensionContext: ExtensionContext?
    ) {
        proxy!!.interceptTestTemplateMethod(invocation, invocationContext, extensionContext)
    }

    @Throws(Throwable::class)
    override fun <T> interceptTestFactoryMethod(
        invocation: InvocationInterceptor.Invocation<T>?,
        invocationContext: ReflectiveInvocationContext<Method>?,
        extensionContext: ExtensionContext?
    ): T? {
        return proxy!!.interceptTestFactoryMethod(invocation, invocationContext, extensionContext)
    }

    @Throws(Throwable::class)
    override fun <T> interceptTestClassConstructor(
        invocation: InvocationInterceptor.Invocation<T>?,
        invocationContext: ReflectiveInvocationContext<Constructor<T>?>?,
        extensionContext: ExtensionContext?
    ): T? {
        return proxy!!.interceptTestClassConstructor(
            invocation,
            invocationContext,
            extensionContext
        )
    }

    companion object {
        private val JUNIT_METHOD_EXECUTION_TIMEOUT_PROPERTY_NAME: String =
            "JunitMethodExecutionTimeout"
    }
}
